rfcs/20190117-tf-module.md

```diff
@@ -250,19 +237,6 @@ class Module(six.with_metaclass(ModuleMetaclass, tf.checkpoint.Checkpointable)):
       followed by variables from all submodules recursively (depth first).
     """
 
-  @property
-  def owned_trainable_variables(self):
-    """Returns a collection of variables that are attributes of the current module.
-
-    See `variables` for a property which returns all variables from the current
-    module and all it's submodules recursively.
-
-    Returns:
-      A collection of variables which are attributes of the current module. Will
-      yield variables inside nested structures (lists etc) but not in other
-      modules.
-    """
-
   @property
   def submodules(self):
```

Comment:
@tomhennigan Is this done depth first or breadth first? I feel like it could be a good thing for `tf.Module` to offer a way to infer the depth of submodules/variables. Without `owned_*`, what would be the best way for a user to do this without re-implementing `tf.Module`'s logic?

Reply:
Breadth first.

So there are two options (frustratingly some recent commits haven't been sync'd to GitHub so I can't just point you at the code!).

Firstly we've made it really easy for subclasses to implement this if the want by exposing `_flatten`:

```python
class MyModule(tf.Module):
  @property
  def owned_submodules(self):
    return tuple(self._flatten(recursive=False, predicate=lambda v: isinstance(v, tf.Module)))
```

I also discussed with brilee@ and pluskid@ earlier about their feedback on variable names. We were aligned that what they both wanted was not really `Variable.name`, but rather the path to all variables (e.g. allowing the same variable instance to be reachable via multiple paths). We're planning to add an option to `_flatten` to allow flattening with paths. So for example in the case where you have:

```python
encoder = Encoder()
decoder = encoder
auto = Auto(encoder, decoder)
```

You can implement `variables_and_paths` in one line with `self._flatten(..., with_path=True)`:

```python
>>> auto.variables_and_paths()
{"encoder.layers[0].w": <tf.Variable ...>,
 "encoder.layers[1].w": <tf.Variable ...>, 
 "decoder.layers[0].w": <tf.Variable ...>,
 "decoder.layers[1].w": <tf.Variable ...>}
```

It seems to me like this is another way of representing depth. WDYT?

Reply:
Actually the CL has been pushed 😄, `_flatten` is in tensorflow/tensorflow@5076adf6

Reply:
I didn't have a particular use case in mind but I believe it is likely for someone to require that type of info (if not simply for human readability). The `with_path` solution is lightweight and should pretty much cover any uses cases related to the topology of the variables/sub-modules. I'm 100% on board, nice job!

Reply:
FYI `with_path` is in tensorflow/tensorflow@1d5fd921 :smile: